### PR TITLE
Change the removal behaviour for overlays

### PR DESCRIPTION
**Pull request by javierarce**
We are changing the current overlay removal behaviour because it's confusing for some users.

Related to this conversation: #2853 

### Here's how it works™

Now we'll remove the overlays in these cases:

1. When the overlay is empty, the cursor is in the text box and the user clicks outside of it.
2. When the overlay is empty, the cursor is in the text box and the user presses the Esc key. 
3. When the overlay is selected and the user presses the backspace.

We won't remove the overlay in this case:

1. When the user removes all the content of the overlay with the backspace.

### CODE REVIEW

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2754/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2806/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2807/))

**Comment by javierarce** [comment]
@viddo || @xavijam: can you take a look to this, please?

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2813/))

File: lib/assets/javascripts/cartodb/table/overlays/annotation.js [hunk 1]
```diff
@@ -169,17 +149,14 @@ cdb.admin.overlays.Annotation = cdb.geo.ui.Annotation.extend({
...
-        if (editable !== "editable") {
-          e.preventDefault();
-          e.stopPropagation();
-
-          this._close();
-        }
+      if (e.keyCode === 8 && (editable || !focus)) {
+        e.preventDefault();
+        e.stopPropagation();
+        this._close();
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb/table/overlays/annotation.js]: this.killEvent(e);

File: lib/assets/javascripts/cartodb/table/overlays/annotation.js [hunk 2]
```diff
@@ -169,17 +149,14 @@ cdb.admin.overlays.Annotation = cdb.geo.ui.Annotation.extend({
 
     if (selected) {
 
-      var editable = this.editModel.get("mode");
+      var editable = this.editModel.get("mode") !== "editable";
+      var focus    = this.$el.find(".overlay_text").is(":focus");
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb/table/overlays/annotation.js]: this.$('.overlay_text')...

File: lib/assets/javascripts/cartodb/table/overlays/annotation.js [hunk 3]
```diff
@@ -240,8 +227,10 @@ cdb.admin.overlays.Annotation = cdb.geo.ui.Annotation.extend({
     var latlng = this.mapView.pixelToLatLon([x , y]);
     extra.latlng = [latlng.lat, latlng.lng];
 
-    this.model.set({ extra: extra });
-    this.model.save();
+    if (this.model.id) { // avoid saving if the model was removed
+      this.model.set({ extra: extra });
+      this.model.save();
+    }
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb/table/overlays/annotation.js]: ```this.model.save({ extra: extra });```  being picky :P

File: lib/assets/javascripts/cartodb/table/overlays/annotation.js [hunk 4]
```diff
@@ -598,13 +587,15 @@ cdb.admin.overlays.Annotation = cdb.geo.ui.Annotation.extend({
...
 
-      setTimeout(function() {
+      //setTimeout(function() {
 
         self.$text.html(text);
-        if (save) self._savePosition();
+        self._savePosition();
 
-      }, 300);
+      //}, 300);
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb/table/overlays/annotation.js]: I'd remove comments if they are not longer needed.

File: lib/assets/javascripts/cartodb/table/overlays/text.js [hunk 5]
```diff
@@ -152,17 +125,14 @@ cdb.admin.overlays.Text = cdb.geo.ui.Text.extend({
...
-        if (editable !== "editable") {
-          e.preventDefault();
-          e.stopPropagation();
-
-          this._close();
-        }
+      if (e.keyCode === 8 && (!editable || !focus)) {
+        e.preventDefault();
+        e.stopPropagation();
+        this._close();
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb/table/overlays/text.js]: Same I said with killEvent...

File: lib/assets/javascripts/cartodb/table/overlays/text.js [hunk 6]
```diff
@@ -471,7 +425,7 @@ cdb.admin.overlays.Text = cdb.geo.ui.Text.extend({
 
     this.editModel.set("mode", "");
 
-    if (text) {
+    if (text && !text.match(/<br ?\/?>/)) { // Firefox adds a <br> tag
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb/table/overlays/text.js]: lol

**Comment by xavijam** [comment]
Non related with these changes, I've seen several ...css({ display: 'inline-block' }). I encourage you to use classes instead of css changes, but as I said, not related with these changes.

**Comment by xavijam** [comment]
Apart from that, :+1:

File: lib/assets/javascripts/cartodb/table/overlays/annotation.js [hunk 7]
```diff
@@ -124,32 +124,12 @@ cdb.admin.overlays.Annotation = cdb.geo.ui.Annotation.extend({
...
-
-    } else {
-
-      this.model.set({ text: value }, { silent: true });
-
+    if (this.timeout) {
+      clearTimeout(this.timeout);
     }
 
+    this.model.set({ text: this.$text.html() }, { silent: true });
```
Comment (viddo) [inline lib/assets/javascripts/cartodb/table/overlays/annotation.js]: why get the HTML markup instead of $text.text()?

**Comment by javierarce** [comment]
Thanks a ton. I'll make the changes.

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2820/))

File: lib/assets/javascripts/cartodb/table/overlays/annotation.js (hunk 7 above)
Comment (javierarce) [inline lib/assets/javascripts/cartodb/table/overlays/annotation.js]: Because I'm silly. Changing it with `text()`. Thanks!

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2821/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2829/))

File: lib/assets/javascripts/cartodb/table/overlays/annotation.js [hunk 8]
```diff
@@ -240,8 +226,9 @@ cdb.admin.overlays.Annotation = cdb.geo.ui.Annotation.extend({
     var latlng = this.mapView.pixelToLatLon([x , y]);
     extra.latlng = [latlng.lat, latlng.lng];
 
-    this.model.set({ extra: extra });
-    this.model.save();
+    if (this.model.id) { // avoid saving if the model was removed
```
Comment (javisantana) [inline lib/assets/javascripts/cartodb/table/overlays/annotation.js]: I think isNew is better here

Comment (xavijam) [inline lib/assets/javascripts/cartodb/table/overlays/annotation.js]: I thought that, but i feel ```isNew``` is not included on cdb.core.Model.

Comment (xavijam) [inline lib/assets/javascripts/cartodb/table/overlays/annotation.js]: Ahh, it is a backbone function. Oooook!

Comment (javierarce) [inline lib/assets/javascripts/cartodb/table/overlays/annotation.js]: I'll change it. Thanks!

Comment (javierarce) [inline lib/assets/javascripts/cartodb/table/overlays/annotation.js]: I'll change it. Thanks!

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2833/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2846/))

**Pull request merged**